src/pipelinedefinitionstatus.hpp

```diff
@@ -0,0 +1,288 @@
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+    StateKeeper handle(const RetireEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+};
+
+struct AvailableState {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::AVAILABLE;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateKeeper handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<AvailableRequiredRevalidation> handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<RetiredState> handle(const RetireEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+};
+
+struct AvailableRequiredRevalidation {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::AVAILABLE_REQUIRED_REVALIDATION;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateChanger<AvailableState> handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
```

Comment:
Are you sure we process UsedModelChangedEvent when in AvailableRequiredRevalidation state? Would it only happen if we have multiple models changing included in the same pipeline?

Reply:
> Would it only happen if we have multiple models changing included in the same pipeline?
Exactly that is the reason why we need to handle multiple modelchanging events. The same is true for LOADING_FAILED_PRECONDITION_REQUIRED_REVALIDATION.

Alternative would be to check if pipeline is already in that state in model/model instance before notifying but that defeats the purpose of keeping , managing changing state logic in PipelineDefinitionStatus implementation.